### PR TITLE
Context implementation

### PR DESCRIPTION
**Pull request by furusiyya**
#75 #82 fixed!

Previously I selected [cli](https://github.com/urfave/cli) for two reasons:
1)- Flags management
2)- It provides context that can travels to all handler functions of application so we can use
	context for sending values and signal across different goroutines.
3)- Flags can be accessed and set at any level of application with the help of context.

While exploring docs of cli I found that we cannot add value in [cli.Context](https://godoc.org/github.com/urfave/cli#Context) structure but it is designed to be only used for carrying context of application flags throughout the hierarchy, mostly used by cli application having a lot flags and sub commands for their management. 
So I decided to use [docopt](https://github.com/docopt/docopt.go) for flag management and native [context](https://golang.org/pkg/context/) package of golang for passing application context to goroutines. For passing flags down to any level of application context can be used because it support for carrying values.

### CODE REVIEW

File: glutton.go [hunk 1]
```diff
@@ -186,6 +198,8 @@ func (g *Glutton) registerHandlers() {
 // Shutdown the packet processor
 func (g *Glutton) Shutdown() (err error) {
 	defer g.logger.Sync()
+	g.cancel() // close all connection
+	time.Sleep(4 * time.Second)
```
Comment (glaslos) [inline glutton.go]: Is there no better way to wait for all connections to be closed?

File: logger.go [hunk 2]
```diff
@@ -8,8 +8,13 @@ import (
 func initLogger(logPath *string, id string, debug *bool) (*zap.Logger, error) {
 
 	cfg := zap.NewProductionConfig()
-	cfg.ErrorOutputPaths = []string{*logPath + "/logger.err"}
-	cfg.OutputPaths = []string{*logPath + "/glutton.log"}
+	if *logPath != "/dev/null" {
```
Comment (glaslos) [inline logger.go]: I'm still not super happy with this solution... Maybe call it log file and error log file? Hard coding the file names seems a bit restrictive and maybe counter-intuitive.

File: protocols.go [hunk 3]
```diff
@@ -1,45 +1,116 @@
...
+			g.logger.Error(fmt.Sprintf("[glutton  ]  error: %v", err))
+		}
+		return
+	case <-done:
+		return
+	}
+}
+
+// Drive child context from parent context with additional value required for sepcific handler
+func (g *Glutton) withTimeout(timeInSeconds uint8) context.Context {
```
Comment (glaslos) [inline protocols.go]: How about `contextWithTimeout`?

File: protocols.go [hunk 4]
```diff
@@ -1,45 +1,116 @@
...
+	limit := time.Duration(timeInSeconds) * time.Second
+	return context.WithValue(g.ctx, "timeout", time.Now().Add(limit))
+}
+
+// updateIdleTime increase connection timeout limit on connection I/O operation
+func (g *Glutton) updateIdleTime(ctx context.Context, conn net.Conn) {
```
Comment (glaslos) [inline protocols.go]: `updateConnectionTimeout` ?

File: protocols.go [hunk 5]
```diff
@@ -1,45 +1,116 @@
...
 	"strings"
+	"time"
 )
 
 // mapProtocolHandlers map protocol handlers to corresponding protocol
 func (g *Glutton) mapProtocolHandlers() {
 
-	g.protocolHandlers["smtp"] = func(conn net.Conn) error {
-		return g.HandleSMTP(conn)
+	g.protocolHandlers["smtp"] = func(conn net.Conn) (err error) {
```
Comment (glaslos) [inline protocols.go]: There is lot's of duplicate code in setting up the connection handlers.

File: logger.go (hunk 2 above)
Comment (furusiyya) [inline logger.go]: I agree. I will fix this.

File: protocols.go (hunk 3 above)
Comment (furusiyya) [inline protocols.go]: sounds better!

File: app/server.go [hunk 6]
```diff
@@ -1,18 +1,30 @@
 package main
 
 import (
-	"flag"
 	"fmt"
 	"os"
 	"os/signal"
 	"sync"
 
+	docopt "github.com/docopt/docopt-go"
```
Comment (furusiyya) [inline app/server.go]: yes

File: glutton.go (hunk 1 above)
Comment (furusiyya) [inline glutton.go]: May be there exist a better but I am unable to find. The only link we have between program and goroutines is context. `cancel()` function signal  routines to abandon their work and does not wait for the work to stop. And in any case if fails then there will be definitely a goroutine leak. May be it is possible in future when we have connection counter so we can keep that counter synchronized with number of goroutines (connections) with help of context and on shutdown we wait until counter goes to zero.

Comment (glaslos) [inline glutton.go]: That argument is good enough for me :) Keep it as it is, maybe add a TODO here or maybe an issue in github.

File: glide.lock [hunk 7]
```diff
@@ -81,6 +83,8 @@ imports:
   version: 645ef00459ed84a119197bfb8d8205042c6df63d
 - name: github.com/satori/go.uuid
   version: b061729afc07e77a8aa4fad0a2fd840958f1942a
+- name: github.com/Sirupsen/logrus
```
Comment (glaslos) [inline glide.lock]: hm, where is that coming from?

Comment (furusiyya) [inline glide.lock]: It is coming from freki.

**Pull request merged**